jwst/badpix_selfcal/badpix_selfcal_step.py

```diff
@@ -0,0 +1,153 @@
+
+
+from ..stpipe import Step
+from . import badpix_selfcal
+import numpy as np
+from jwst import datamodels as dm
+from jwst.master_background.master_background_step import split_container
+
+import logging
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+__all__ = ["BadpixSelfcalStep"]
+
+
+class BadpixSelfcalStep(Step):
+    """
+    BadpixSelfcalStep: Flags residual artifacts as bad pixels in the DQ array
+    of a JWST exposure using a median filter and percentile cutoffs.
+
+    All input exposures in the association file (or manually-provided bkg_list) are combined
+    into a single background model using a MIN operation. The bad pixels are then identified
+    using a median filter and percentile cutoffs, and applied to the science data by setting
+    the flagged pixels to NaN and the DQ flag to DO_NOT_USE + OTHER_BAD_PIXEL.
+    """
+
+    class_alias = "badpix_selfcal"
+    bkg_suffix = "badpix_selfcal"
+
+    spec = """
+    flagfrac = float(default=0.001)  #fraction of pixels to flag on each of low and high end
+    kernel_size = integer(default=15)  #size of kernel for median filter
+    force_single = boolean(default=False)  #force single input exposure
+    skip = boolean(default=True)
+    """
+
+    def process(self, input, selfcal_list=[]):
+        """
+        Flag residual artifacts as bad pixels in the DQ array of a JWST exposure
+
+        Parameters
+        ----------
+        input: JWST data model or association
+            input science data to be corrected
+
+        selfcal_list: list of ImageModels or filenames to use for selfcal
+
+        Returns
+        -------
+        output: JWST data model or association
+            data model with CRs flagged
+
+        Notes
+        -----
+        If selfcal_list is specified manually, it overrides any non-science exposures
+        in the association file.
+        If selfcal_list is set to None and an association file is read in, all exposures in the
+        association file, including science, background, and selfcal exposures,
+        are included in the MIN frame from which outliers are detected.
+        If selfcal_list is set to None and input is a single science exposure, the step will
+        be skipped with a warning unless the force_single parameter is set True.
+        In that case, the input exposure will be used as the sole background exposure,
+        i.e., true self-calibration.
+        """
+        input_sci, selfcal_list, bkg_list_asn = _parse_inputs(input, selfcal_list)
+
+        # ensure that there are background exposures to use, otherwise skip the step
+        # unless forced
+        if (len(selfcal_list) == 0) and (not self.force_single):
+            log.warning("No background exposures provided for self-calibration. Skipping step.")
+            self.record_step_status(input_sci, "badpix_selfcal", success=False)
```

Comment:
This function will currently set skip=True for the step. It's probably better to just directly set the calibration step status to SKIPPED.

Reply:
How does the change you are proposing interact with [this ticket](https://jira.stsci.edu/browse/JP-3631)?  And, related, what is the point of having a method called `record_step_status()` if it should not be used?

Reply:
That ticket is why I mentioned it.  It's preferable not to set step.skip = True from inside a pipeline step, because it can have unintended consequences.  It's fine to call record_step_status with success=True, but until we get it updated, best not to call it with success=False.  Or, send in a fix for the function and call it in both places here!

Reply:
You're saying the only way skipping a step should happen is if it's specified in the inputs or by a reference file, is that right?  In that case, I assume I should just make this throw an exception?

Reply:
Setting the skip attribute in the step should only happen at the pipeline level, if the step is truly inappropriate to run on the input data (e.g. an instrument mode mismatch), or else if the user directly specified it.  

It's still okay to skip a step when some internal condition is encountered!  You can just warn, set the SKIPPED status, and return.  You just don't want to set it in the step attribute, because that would still skip the step if some other input data was passed, even if it might otherwise meet your internal criteria.